feat: allow updates to spec.leaderWorkerTemplate.size

synthe102 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature
/kind api-change

#### What this PR does / why we need it

This PR implements KEP 552 by introducing a new field: `spec.leaderWorkerTemplate.resizePolicy`. This field defaults to `None` which keeps the current behavior of not allowing updates to the `size` field. If the value is set to `Recreate`, updates to `spec.leaderWorkerTemplate.size` will be allowed.

#### Which issue(s) this PR fixes
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #552 

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Introduce `resizePolicy` to control updates to the `size` field.
```

kerthcet commented:
/ok-to-test

kerthcet commented on api/leaderworkerset/v1/leaderworkerset_types.go:
```diff
@@ -169,6 +176,14 @@ type LeaderWorkerTemplate struct {
 	// +optional
 	SubGroupPolicy *SubGroupPolicy `json:"subGroupPolicy,omitempty"`
+
+	// ResizePolicy defines how to handle worker size updates.
+	// None doesn't allow to update spec.leaderWorkerTemplate.size.
+	// Recreate updates the size and restarts existing pods.
```
```suggestion
	// ResizePolicy defines how to handle group size updates:
	// - None: indicates the `spec.leaderWorkerTemplate.size` is immutable.
	// - Recreate: indicates recreating the group Pods to reshape the size.
```

kerthcet commented on test/testutils/util.go:
```diff
@@ -392,6 +392,21 @@ func HasLWSEnvVarsPopulated(pod corev1.Pod) bool {
 	return hasAllEnvVarPopulated(pod, []string{leaderworkerset.LwsLeaderAddress, leaderworkerset.LwsGroupSize, leaderworkerset.LwsWorkerIndex})
 }
 
+func CheckSizeAnnotation(pod corev1.Pod, size int) error {
```
Can we make this move general like`CheckAnnotation(pood, key, value)`

synthe102 replied:
Sure !

kerthcet reviewed (commented):
To make sure I understand the whole process: once we change the size, the leader sts will update the size value in the pod template annotation, which will trigger the rolling update of leader pods, then the worker pods, right?

Would you like to take another look? @Edwinhr716

Edwinhr716 commented:
> once we change the size, the leader sts will update the size value in the pod template annotation, which will trigger the rolling update of leader pods, then the worker pods

Yes that would happen, but using this current implementation we don't trigger a rolling update at the LWS controller level.

Edwinhr716 commented:
LGTM overall. @synthe102 did you do manual testing to confirm that a rolling update is triggered at the LWS level? If you check the events of the LWS object during it and it shows a RollingUpdateInProgress event, then it that would confirm that it was at the LWS level

kerthcet commented:
> d it shows a RollingUpdateInProgress eve

Let's add a `ValidateEvent` then, then we can merge I think.

yankay commented on config/crd/bases/leaderworkerset.x-k8s.io_leaderworkersets.yaml:
```diff
@@ -8184,6 +8184,16 @@ spec:
+                    - None
+                    - Recreate
+                    type: string
                   restartPolicy:
```
Do the CRDs in the charts also need to be modified :-)

ref: https://github.com/kubernetes-sigs/lws/blob/main/charts/lws/templates/crds/leaderworkerset.x-k8s.io_leaderworkersets.yaml

kerthcet commented:
> Just to be sure, you want add a new Event Reason `Validate` that will be emitted when a resize occurs ? I'm not sure I understand where this event will be used, could you clarify ?

Sorry, I mean we can add a event validation like this: https://github.com/kubernetes-sigs/lws/blob/82b36869d3f774bd6b022353b434e813de84296a/test/integration/controllers/leaderworkerset_test.go#L205-L207

so solve Edwin's concern.

ahg-g commented on test/integration/webhooks/leaderworkerset_test.go:
```diff
@@ -261,6 +261,24 @@ var _ = ginkgo.Describe("leaderworkerset defaulting, creation and update", func(
 			},
 			lwsCreationShouldFail: true,
 		}),
+		ginkgo.Entry("update to size when resizePolicy is None should fail", &testValidationCase{
```
This is only testing the validation logic. Can we add an integration test to verify that the controller actually applies this change to the statefulsets?

kerthcet reviewed (commented):
I think we still miss three parts right now:
- change the `resizePolicy ` field to a pointer
- add an integration test to verify that the size is indeed updated
- add the rolling update event validation

All of these three are very simple ones I think, so we're very close now.  :)

synthe102 commented:
> I think we still miss three parts right now:
> 
>     * change the `resizePolicy ` field to a pointer
> 
>     * add an integration test to verify that the size is indeed updated
> 
>     * add the rolling update event validation
> 
> 
> All of these three are very simple ones I think, so we're very close now. :)

I think we should be good. I changed the `ResizePolicy` to a pointer, added an integration test to verify that the size is updated, and check that we have a rolling update event.

kerthcet commented:
Will take a look later today, Thanks @synthe102

kerthcet commented on test/integration/controllers/leaderworkerset_test.go:
```diff
@@ -1971,6 +1971,41 @@ var _ = ginkgo.Describe("LeaderWorkerSet controller", func() {
+						testing.UpdateSize(ctx, k8sClient, lws, 3)
+						testing.SetPodGroupsToReady(ctx, k8sClient, lws, 2)
+					},
+					checkLWSState: func(lws *leaderworkerset.LeaderWorkerSet) {
```
Nit: let's verify that before the size is 2, after update the size increased to 3, I think we missed that part.

synthe102 replied:
I would say that `ExpectValidWorkersStatefulSets` being called before and after the resize takes care of that ? [Here](https://github.com/kubernetes-sigs/lws/blob/8d65262f630745031013318c8861d22fb892e738/test/testutils/validators.go#L286-L288) and [here](https://github.com/kubernetes-sigs/lws/blob/8d65262f630745031013318c8861d22fb892e738/test/testutils/validators.go#L286-L288) for the size annotation.

kerthcet replied:
Although not that straightforward, but fair enough, since we assume that the update is succeeded.

kerthcet reviewed (commented):
/approve

Generally LGTM. Only one nit from my side.

kerthcet commented:
/lgtm
Let's merge now since we addressed all the comments and other PRs may hit code conflicts with this one, we're close to the deadline, then can save some time for them to rebase. 

Thanks @synthe102 for the patient and efforts, it's great to see this feature to be merged in v0.7.0, also glad to see MistralAI as one of the users, maybe you're happy to join the list: https://lws.sigs.k8s.io/docs/adoption/

JesseStutler commented:
<img width="1981" height="446" alt="image" src="https://github.com/user-attachments/assets/768220ec-aba6-4854-bcab-eda851ad7db1" />
@synthe102 @kerthcet Hi, I noticed that the `resize shoud update the size of the replicas` integration test failed in my environment. Is it a bug or I missed something to configure in my environment? But other 44 cases can pass

JesseStutler commented:
> Do you mean the main branch? I checked locally, all tests passed.

Yes main branch, maybe it's because my environment's fault, never mind 😄

ahg-g commented:
Looking at this again, why do we need an api to select between two modes? why not just always allow resize? this is a backward compatible change since this is simply relaxing validation.

ahg-g commented:
> And what if people don't want to change the size?

Then they shouldn't update it :)

ahg-g commented:
Relaxing validation is generally a backward compatible change.

kerthcet commented:
> but from the API perspective, it's indeed backward compatible

Yes I agree that relaxing the validation is backward compatible. So you strongly prefer to remove the resizePolicy, right? @ahg-g
